### PR TITLE
add workaround on extract method on https://github.com/dotnet/roslyn/…

### PR DESCRIPTION
**Pull request by heejaechang**
…issues/30548

### CODE REVIEW

**Comment by heejaechang** [comment]
root issue is tracked here - https://github.com/dotnet/roslyn/issues/30548

**Comment by heejaechang** [comment]
@agocke added unit test in compiler side for the root cause - SymbolInDataFlowInButNotInReadInside

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs [hunk 1]
```diff
@@ -6314,6 +6312,54 @@ void M()
...
+}");
+
+            var dataFlowAnalysisResults = analysisResults.Item2;
+
+            Assert.Equal(null, GetSymbolNamesJoined(dataFlowAnalysisResults.VariablesDeclared));
+            Assert.Equal(null, GetSymbolNamesJoined(dataFlowAnalysisResults.AlwaysAssigned));
+            Assert.Equal(null, GetSymbolNamesJoined(dataFlowAnalysisResults.Captured));
+            Assert.Equal(null, GetSymbolNamesJoined(dataFlowAnalysisResults.CapturedInside));
+            Assert.Equal(null, GetSymbolNamesJoined(dataFlowAnalysisResults.CapturedOutside));
+            Assert.Equal("test, a", GetSymbolNamesJoined(dataFlowAnalysisResults.DataFlowsIn));
```
Comment (heejaechang) [inline src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs]: right now, the test captures current behavior. either "test" should be removed from "DataFlowsIn" or "test" should be added to "ReadInside"

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs]: @heejaechang can you explain this?  Why shoudl that be the case?

File: src/Features/Core/Portable/ExtractMethod/ExtractMethodMatrix.cs [hunk 2]
```diff
@@ -158,6 +158,10 @@ private static void BuildMatrix()
             s_matrix.Add(new Key(dataFlowIn: true, dataFlowOut: false, alwaysAssigned: true, variableDeclared: false, readInside: true, writtenInside: true, readOutside: false, writtenOutside: true), VariableStyle.InputOnly);
             s_matrix.Add(new Key(dataFlowIn: true, dataFlowOut: false, alwaysAssigned: true, variableDeclared: false, readInside: true, writtenInside: true, readOutside: true, writtenOutside: false), VariableStyle.InputOnly);
             s_matrix.Add(new Key(dataFlowIn: true, dataFlowOut: false, alwaysAssigned: true, variableDeclared: false, readInside: true, writtenInside: true, readOutside: true, writtenOutside: true), VariableStyle.InputOnly);
+
+            // workaround for the bug(https://github.com/dotnet/roslyn/issues/30548)
+            s_matrix.Add(new Key(dataFlowIn: true, dataFlowOut: true, alwaysAssigned: false, variableDeclared: false, readInside: false, writtenInside: true, readOutside: true, writtenOutside: true), VariableStyle.Ref);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ExtractMethod/ExtractMethodMatrix.cs]: 1. can you explain how this is a bug?
2. this fix seems insufficient as it doesn't seem like it will prevent goldbars for other scenarios.  can we not crash in that case, but instead just send a telemetry report about a case we're not handling?

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs (hunk 1 above)
Comment (heejaechang) [reply src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs]: probably compiler team is a better one to explain why that's the case. 

my simplified version is this
data flow in means someone consumed the data assigned outside, and there is no one consuming the data.

that's all I have. so if you can't understand, just ask compiler team. I dont have anything better.

for exact definition of all those APIs, probably @agocke or @gafter better person to give you ones you want.

File: src/Features/Core/Portable/ExtractMethod/ExtractMethodMatrix.cs (hunk 2 above)
Comment (heejaechang) [reply src/Features/Core/Portable/ExtractMethod/ExtractMethodMatrix.cs]: sure. you can use "crash" any way you want. I just wanted to make sure others not misunderstood it as the "crash" as we use it - process going away since it is just infobar not process going away.
...

I already wrote why it is a bug in compiler issue and this PR. so keep asking me why it is a bug will just go in a circle. so again, it would be better to direct that question to compiler team.
..

I was planning to fix C7 and if that work involved changing info bar behavior I would have changed that as well, but that is separate work. so I don't want to mix two. I opened a tracking issue for it. 

basically, I dont want to get into some kind of discussion on info bar experience blocking this one from getting any progress. I have other things to do.

**Comment by heejaechang** [comment]
retest windows_release_vs-integration_prtest please

**Comment by heejaechang** [comment]
@dotnet/roslyn-ide @mavasani can you take a look?

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs [hunk 3]
```diff
@@ -6314,6 +6312,54 @@ void M()
             Assert.Equal("this, x", GetSymbolNamesJoined(dataFlowAnalysisResults.WrittenOutside));
         }
 
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/30548")]
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs]: why skip?

Comment (heejaechang) [reply src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs]: because I wasn't sure what is expected behavior. either "test" shouldn't be in data flow in or "test" should be in read inside.

test is added just to reduce labor for compiler team.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs]: It would be better ot be not skipped.  If the test is correct, there's nothing that needs to chane.  if the test is incorrect, it's just docmenting existing behavior and will break if the compiler thinks that behavior needs to change.  otherwise, it's just another skipped test sitting around that never gets run or looked at again.

Comment (heejaechang) [reply src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs]: I have added tests for compiler team this way many times. never saw they left it as skipped once they fixed the bug.

**Comment by heejaechang** [comment]
retest windows_debug_spanish_unit32_prtest please

**Comment by heejaechang** [comment]
Spanish unit32 never pass the test.

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure spanish unit 32 seems not doing anything?
